### PR TITLE
Fix NLL issue 50716 and add a regression test.

### PR DESCRIPTION
**Pull request by vakaras**
Fix for NLL issue #50716.

r? @nikomatsakis

### CODE REVIEW

**Review by pnkfelix — COMMENTED** [review]
I think the test needs to either be moved to the `ui/` directory (and a corresponding `.stderr` file is then also added to the PR), *or* it needs an explicit `#![feature(nll)]`. As it stands right now I don’t think it’s actually testing the code you’ve added.

**Comment by vakaras** [comment]
Good catch @pnkfelix! I have updated the test.

File: src/test/compile-fail/nll/issue-50716.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+trait A {
+    type X: ?Sized;
+}
+
+fn foo<'a, T: 'static>(s: Box<<&'a T as A>::X>)
+where
+    for<'b> &'b T: A,
+    <&'static T as A>::X: Sized
+{
+    let _x = *s; //~ ERROR free region `'a` does not outlive free region `'static`
```
Comment (nikomatsakis) [inline src/test/compile-fail/nll/issue-50716.rs]: the `compile-fail` test directory is basically deprecated... I'd prefer this to be in `src/test/ui/nll/` =)

**Review by nikomatsakis — DISMISSED** [review]
Sorry for ping-ponging!

**Comment by vakaras** [comment]
@pnkfelix Thanks!

**Comment by vakaras** [comment]
@pnkfelix @nikomatsakis The error we got on bors looks very similar to the one I was getting during the Rust Fest:
```
error: internal compiler error: broken MIR in NodeId(3244) (""): errors selecting obligation: [FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<Q as std::marker::Sized>)),depth=1),Unimplemented)]
  --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\linked-hash-map-0.3.0\src\lib.rs:86:1
   |
86 | struct Qey<Q: ?Sized>(Q);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
error: aborting due to previous error
```
Do you have any idea how to investigate it? `./x.py build` and `./x.py test` works on my machine without any issues.

**Comment by nikomatsakis** [comment]
@michaelwoerister do we use incremental on bors ? presumably not...

**Comment by nikomatsakis** [comment]
(No reason to blame incremental per se)

**Comment by nikomatsakis** [comment]
@vakaras I don't know but it does seem like it could be related to this PR — after all, you added an obligation to prove `Sized`, and it is failing without proving that.

**Comment by TimNN** [comment]
@vakaras: Do you think you'll be able to investigate this issue? It looks like compiling `linked-hash-map@0.3.0` causes the issue (This happens in a `cargotest`, which isn't run by `./x.py test` by default, I believe. Can you reproduce the issue if you build a local compiler and then build the crate manually?

**Comment by vakaras** [comment]
@TimNN  I am not sure I understand well enough what is happening here to investigate this issue properly. However, I will try at least to reproduce the error on my machine.

**Comment by vakaras** [comment]
@TimNN Thanks! You were right, the following steps reproduce the issue:
```bash
git clone https://github.com/vakaras/rust.git -b issue-50716 issue-50716
cd issue-50716/
./x.py build
./x.py test src/tools/cargotest src/tools/cargo
```
I will try to investigate this issue.

**Comment by kennytm** [comment]
@bors r-

Wrongly requeued

**Comment by vakaras** [comment]
@nikomatsakis The minimal not working example (taken from `linked-hash-map`):
```rust
struct Qey<Q: ?Sized>(Q);
```
Do you have any idea what could be wrong here?


Full repro steps:
```
git clone https://github.com/vakaras/rust.git -b issue-50716 issue-50716
cd issue-50716/
./x.py build
./x.py test src/tools/cargotest src/tools/cargo
rustup toolchain link issue-50716 $PWD/build/x86_64-unknown-linux-gnu/stage2/
echo 'struct Qey<Q: ?Sized>(Q);' > code.rs
rustc +issue-50716 --crate-type lib code.rs
warning: struct is never used: `Qey`
 --> code.rs:1:1
  |
1 | struct Qey<Q: ?Sized>(Q);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default

error: internal compiler error: broken MIR in DefId(0/1:9 ~ code[8787]::Qey[0]::{{constructor}}[0]) (""): errors selecting obligation: [FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<Q as std::marker::Sized>)),depth=1),Unimplemented)]
 --> code.rs:1:1
  |
1 | struct Qey<Q: ?Sized>(Q);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

**Comment by nikomatsakis** [comment]
@vakaras I guess that the problem comes from the "auto-generated" MIR for the constructor of `Qey`, which looks like this:

```rust
fn Qey(_1: Q) -> Qey<Q>{
    let mut _0: Qey<Q>;                  // return place

    bb0: {                              
        (_0.0: Q) = move _1;             // bb0[0]: scope 0 at src/main.rs:1:1: 1:26
        return;                          // bb0[1]: scope 0 at src/main.rs:1:1: 1:26
    }
}
```

You can see that this does `_0.0 = ...` which -- indeed -- is not entirely valid, since we do not know that `Q: Sized`. (In fact, we *do* know this, because we would not allow a call `Q(bar)` unless `bar` were sized, but the current MIR doesn't represent that anyhow.)

I can think of two fixes.

One might be to skip borrowck for such generated code -- in fact, we should probably do that just for performance reasons. 

The other would be to modify the generated code to carry an extended environment. While more correct, I'm not inclined to do that just because it's complicated.

To skip this we would probably want to modify `do_mir_borrowck`. We already have a "shortcircuit check" that just looks to see if MIR borrow check is enabled:

https://github.com/rust-lang/rust/blob/2a1c4eec40527de45b9d9b81672c8b9220d554fc/src/librustc_mir/borrow_check/mod.rs#L79-L84

Maybe we should add some code that checks if the variable `def_id` (which uniquely identifies the code to be type-checked) comes from a "struct constructor" (which is what we call the auto-geneated tuple struct constructors). To do that, I would suggest adding a fn to `TyCtxt` similar to the existing `is_closure`:

https://github.com/rust-lang/rust/blob/2a1c4eec40527de45b9d9b81672c8b9220d554fc/src/librustc/ty/util.rs#L521-L523

This might look like:

```rust
/// True if this def-id refers to the implicit constructor for a tuple struct like `struct Foo(u32)`
pub fn is_struct_constructor(self, def_id: DefId) -> bool {
        self.def_key(def_id).disambiguated_data.data == DefPathData::StructCtor
}
```

(What this is doing is looking up, via `def_key`, some information about what the def-id refers to.)

Then we can modify MIR borrowck to return early if `tcx.is_struct_constructor(def_id)` is true (with an explanatory comment citing this example!)

Also, let's add this as a test case =)

**Comment by vakaras** [comment]
@nikomatsakis I have tried to implement your suggestions [here](https://github.com/vakaras/rust/commit/292aa68e845a9389b59effd5bad0c1ad328bd220). However, it does not seem to fix the issue and for some reason affects the reported error spans in two tests.

**Comment by vakaras** [comment]
I think I finally understood why it still crashes. The fix in the pull request fixes only when NLL is used. However, when the old borrow checker is used, the MIR is still type checked by this code:
https://github.com/rust-lang/rust/blob/c20824323cf7ed6ad95cb8d7b780a7934927a753/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L1708-L1740
I will try to implement the same early return for this function too.

File: src/librustc_mir/borrow_check/mod.rs [hunk 2]
```diff
@@ -76,7 +76,37 @@ fn mir_borrowck<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> BorrowC
...
+        //     bb0: {
+        //         (_0.0: Q) = move _1;             // bb0[0]: scope 0 at src/main.rs:1:1: 1:26
+        //         return;                          // bb0[1]: scope 0 at src/main.rs:1:1: 1:26
+        //     }
+        // }
+        // ```
+        // The problem here is that `(_0.0: Q) = move _1;` is valid only if `Q` is
+        // of statically known size, which is not known to be true because of the
+        // `Q: ?Sized` constraint. However, it is true because the constructor can be
+        // called only when `Q` is of statically known size.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: Nice comment =)

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**